### PR TITLE
[v1.0] Move Domain-Specific Model Builder into DENAS

### PR DESCRIPTION
**Pull request by tianyil1**
This PR is updated with the followings:

- Refine the e2eaiok/common model builder structure, and move denas-specific model builder into denas folder
- Refine the e2eaiok/common data builder structure

### CODE REVIEW

File: e2eAIOK/DeNas/asr/model_builder_denas_asr.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import torch
+from e2eAIOK.common.trainer.model_builder import ModelBuilder
+from e2eAIOK.DeNas.asr.lib.convolution import ConvolutionFrontEnd
+from e2eAIOK.DeNas.module.asr.linear import Linear
+from e2eAIOK.DeNas.asr.data.processing.features import InputNormalization
+from e2eAIOK.DeNas.module.asr.utils import gen_transformer
+from e2eAIOK.DeNas.utils import decode_arch_tuple
+
+class ModelBuilderASR(ModelBuilder):
```
Comment (XinyaoWa) [inline e2eAIOK/DeNas/asr/model_builder_denas_asr.py:9]: change to:
    class ModelBuilderASRDeNas(ModelBuilderASR):

File: e2eAIOK/DeNas/cv/model_builder_denas_cv.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+from e2eAIOK.DeNas.utils import decode_arch_tuple
+from e2eAIOK.DeNas.cv.third_party.ZenNet import DeMainNet
+from e2eAIOK.common.trainer.model_builder import ModelBuilder
+from e2eAIOK.DeNas.cv.supernet_transformer import Vision_TransformerSuper
+
+class ModelBuilderCV(ModelBuilder):
```
Comment (XinyaoWa) [inline e2eAIOK/DeNas/cv/model_builder_denas_cv.py:6]: change to:
    class ModelBuilderCVDeNas(ModelBuilderCV):

File: e2eAIOK/DeNas/nlp/model_builder_denas_nlp.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import ast
+import os
+import logging
+import torch
+
+from e2eAIOK.common.trainer.model_builder import ModelBuilder
+from e2eAIOK.DeNas.nlp.supernet_bert import SuperBertForQuestionAnswering, BertConfig
+from e2eAIOK.DeNas.nlp.utils import decode_arch
+
+class ModelBuilderNLP(ModelBuilder):
```
Comment (XinyaoWa) [inline e2eAIOK/DeNas/nlp/model_builder_denas_nlp.py:10]: change to:
    class ModelBuilderNLPDeNas(ModelBuilderNLP):

File: e2eAIOK/common/trainer/model/model_builder_asr.py [hunk 4]
```diff
@@ -1,54 +1 @@
...
-        modules["CNN"] = cnn
-        modules["Transformer"] = transformer
-        modules["seq_lin"] = seq_lin
-        modules["ctc_lin"] = ctc_lin
-        modules["normalize"] = normalize
-        model = torch.nn.ModuleDict(modules)
-        
-        return model
\ No newline at end of file
+# Todo: Add common asr model features into this module in the future
```
Comment (XinyaoWa) [inline e2eAIOK/common/trainer/model/model_builder_asr.py:1]: Keep following code:

import torch
from e2eAIOK.common.trainer.model_builder import ModelBuilder

class ModelBuilderASR(ModelBuilder):
    def __init__(self, cfg):
        super().__init__(cfg)

File: e2eAIOK/common/trainer/model/model_builder_cv.py [hunk 5]
```diff
@@ -1,40 +1 @@
...
-            model_config['layer_num'] = depth
-            model_config['mlp_ratio'] = mlp_ratio
-            model_config['num_heads'] = num_heads
-            model_config['embed_dim'] = [embed_dim]*depth
-            n_parameters = model.get_sampled_params_numel(model_config)
-            print("model parameters size: {}".format(n_parameters))
-        return model
-        
\ No newline at end of file
+# Todo: Add common cv model features into this module in the future
```
Comment (XinyaoWa) [inline e2eAIOK/common/trainer/model/model_builder_cv.py:1]: Keep following code:

import torch
from e2eAIOK.common.trainer.model_builder import ModelBuilder

class ModelBuilderCV(ModelBuilder):
    def __init__(self, cfg):
        super().__init__(cfg)

File: e2eAIOK/common/trainer/model/model_builder_nlp.py [hunk 6]
```diff
@@ -1,24 +1 @@
...
-        device = torch.device("cuda" if torch.cuda.is_available() and not self.args.no_cuda else "cpu")
-        model.to(device)
-        subbert_config = decode_arch(self.cfg.best_model_structure)
-        model.module.set_sample_config(subbert_config) if hasattr(model, 'module') else model.set_sample_config(subbert_config)
-        size = model.module.calc_sampled_param_num() if hasattr(model, 'module') else model.calc_sampled_param_num()
-        print("architecture: {}".format(subbert_config))
-        print('Total parameters: {}'.format(size))
-        return model
\ No newline at end of file
+# Todo: Add common nlp model features into this module in the future
```
Comment (XinyaoWa) [inline e2eAIOK/common/trainer/model/model_builder_nlp.py:1]: Keep following code:

import torch
from e2eAIOK.common.trainer.model_builder import ModelBuilder

class ModelBuilderNLP(ModelBuilder):
    def __init__(self, cfg):
        super().__init__(cfg)

Comment (tianyil1) [reply e2eAIOK/common/trainer/model/model_builder_nlp.py]: I have fixed the above comments. Please check it.

**Comment by tianyil1** [comment]
Please help review this PR. Thx. @xuechendi

**Review by XinyaoWa — APPROVED** [review]

File: e2eAIOK/DeNas/train.py [hunk 7]
```diff
@@ -9,14 +9,14 @@
 import e2eAIOK.common.trainer.utils.extend_distributed as ext_dist
 from e2eAIOK.common.trainer.torch_trainer import TorchTrainer 
 import e2eAIOK.common.trainer.utils.utils as utils
-from e2eAIOK.common.trainer.model.model_builder_asr import ModelBuilderASR
-from e2eAIOK.common.trainer.model.model_builder_cv import ModelBuilderCV
-from e2eAIOK.common.trainer.model.model_builder_nlp import ModelBuilderNLP
-from e2eAIOK.common.trainer.data.data_builder_librispeech import DataBuilderLibriSpeech
-from e2eAIOK.common.trainer.data.data_builder_cifar import DataBuilderCIFAR
+from asr.model_builder_denas_asr import ModelBuilderASRDeNas
```
Comment (xuechendi) [inline e2eAIOK/DeNas/train.py:12]: please use full path here instead of relative path
change `asr.model_builder_denas_asr` to `e2eAIOK..denas.asr.model_builder_denas_asr`

File: e2eAIOK/DeNas/train.py [hunk 8]
```diff
@@ -9,14 +9,14 @@
...
+from cv.model_builder_denas_cv import ModelBuilderCVDeNas
+from nlp.model_builder_denas_nlp import ModelBuilderNLPDeNas
```
Comment (xuechendi) [inline e2eAIOK/DeNas/train.py:14]: same as above

**Comment by xuechendi** [comment]
@tianyil1 , codes looks good to me, one comment is you need to use full package path instead relative path
Please submit an extra commit to trigger CI/CD, once it past, I'll merge it.

**Comment by tianyil1** [comment]
> @tianyil1 , codes looks good to me, one comment is you need to use full package path instead relative path Please submit an extra commit to trigger CI/CD, once it past, I'll merge it.

I have fixed the above issues. Please check it. @xuechendi

**Comment by Peach-He** [comment]
> @tianyil1 , codes looks good to me, one comment is you need to use full package path instead relative path Please submit an extra commit to trigger CI/CD, once it past, I'll merge it.

@xuechendi please help to add jenkins test labels to trigger CI/CD

**Comment by xuechendi** [comment]
@tianyil1 , there is a conflict you need to resolve, I think this is related to CNN recent change, please check with @csdingbin to align

**Comment by tianyil1** [comment]
> @tianyil1 , there is a conflict you need to resolve, I think this is related to CNN recent change, please check with @csdingbin to align

I have fixed the above conflict. Please check it. Thx.

**Review by xuechendi — APPROVED** [review]
approved
